### PR TITLE
Drop support for legacy Python 2.7

### LINKED ISSUE #434: Drop legacy python support
I'd like to open a discussion about dropping support for legacy python to make our lives easier.
I came across multiple issues during development or #411 especially during CI tests, most issues with tests were simply fixed by testing on python versions `3.5, 3.6, 3.7`.

Moreover `requests` and other packages are dropping legacy python as well (more info at https://python3statement.org)

On the top of that `requests3` are in development. 

I think project can benefit from that and we can maintain less code.

What do you think about it?

### PR DESCRIPTION
**Pull request by hugovk**
Fixes #434.

Python 2.7 reaches EOL on 2020-01-01. Many projects are pledging to drop support before or during 2020: https://python3statement.org/

It's also relatively little used. Here's the pip installs for vcrpy from PyPI for May 2019:

| category | percent | downloads |
|----------|--------:|----------:|
|      3.6 |  37.19% |    65,506 |
|      3.7 |  26.36% |    46,432 |
|      2.7 |  21.83% |    38,445 |
|      3.5 |  11.53% |    20,303 |
|      3.4 |   2.48% |     4,361 |
| null     |   0.31% |       553 |
|      3.8 |   0.29% |       503 |
|      3.3 |   0.01% |        14 |
| Total    |         |   176,117 |

Source: `pypistats python_minor vcrpy --last-month # pip install pypistats`

There's one bit of `six` remaining in this PR, what's the best way to replace this?

```python
    def test_case(self, predicate=None):
        predicate = predicate or self.is_test_method
        return six.with_metaclass(auto_decorate(self.use_cassette, predicate))
```

### CODE REVIEW

**Comment by IvanMalison** [comment]
hmmm. I think this is a @kevin1024 decision. I can see arguments on both sides.

I would squash commits for this change so we get a nice clean changelog

**Comment by hugovk** [comment]
Happy to squash, or this can also be done at merge time:

image

**Comment by IvanMalison** [comment]
@hugovk fix conflicts

**Comment by hugovk** [comment]
Resolved!

**Comment by neozenith** [comment]
Looks like new conflicts with master. 😢

Looks soo good seeing all the python2 code getting gutted. 

Dropping support would be a breaking change making this inclusion a v3 upgrade to vcr. (FYI @arthurHamon2)

Let’s iterate on this PR and get it merged by the end of the week before any more conflicts crop up.

**Comment by arthurHamon2** [comment]
I'm not sure we have to rush it ! This is a big breaking change and the downloading stats show that 21% still come from Python 2. Maybe it would be wise to release a last version 2.X with the bug fixes and warn that Python 2 will be removed in the future 3.0 release. 
@kevin1024 what do you think ?

**Comment by neozenith** [comment]
Great point @arthurHamon2 and that'd add more value to this project being empathetic of v2.7 users until the last gasp as they are often there not by choice. For the `v2.1.x` we should slip in a deprecation warning.

I'd like to close out @hugovk 's work so they do not have to keep rebasing with master until we release a v3.x

Is this where other projects potentially create `v2.1.x`, `v2.2.x`, `v3.0.x` branches to accumulate changes for different releases?

https://producingoss.com/en/release-lines.html

This article had some good discussion on the topic.

File: docs/contributing.rst [hunk 1]
```diff
@@ -11,13 +11,10 @@ yourself using `py.test <http://pytest.org/>`__ and
 all environments VCR.py supports. The test suite is pretty big and slow,
 but you can tell tox to only run specific tests like this::
 
-    tox -e {pyNN}-{HTTP_LIBRARY} -- <pytest flags passed through>
-
-    tox -e py27-requests -- -v -k "'test_status_code or test_gzip'"
-    tox -e py37-requests -- -v --last-failed
+    tox -e py37requests -- -v -k "'test_status_code or test_gzip'"
```
[the remark this answers is not in the document]
Comment (hugovk) [reply docs/contributing.rst]: Thanks, will fix that. Will also add back the other line but change `27` to `36` as it shows some useful options:

```
    tox -e py36-requests -- -v -k "'test_status_code or test_gzip'"
    tox -e py37-requests -- -v --last-failed
```

**Comment by hugovk** [comment]
Some projects have already dropped Python 2.7 to reduce maintenance burden. Others are waiting until the EOL before dropping. Others are dropping 2.7 now/soon but creating a 2.7 maintenance branch, and keeping that until EOL, or only accepting community contributions after that.

Some examples:

* https://www.numpy.org/neps/nep-0014-dropping-python2.7-proposal.html
* https://docs.pytest.org/en/latest/py27-py34-deprecation.html
* https://pillow.readthedocs.io/en/stable/deprecations.html#python-2-7
* https://docs.djangoproject.com/en/dev/faq/install/#what-python-version-should-i-use-with-django

**Comment by hugovk** [comment]
Rebased and updated!

**Comment by neozenith** [comment]
Hi @hugovk I haven't forgotten this PR. I've assigned it to the v3 milestone.

There are some bug reports coming in for v2.1.0 that I'd like to address and also add Python3.8 support into our TravisCI matrix targetted for v2.2.0 hopefully in September before it's release in October.

Then pull the trigger on v3.0 dropping python 2 support.

Just wanted to give you an update and thanks for your patience. I'm looking forward to culling that maintenance burden.

**Comment by hugovk** [comment]
Thanks, sounds good!

**Comment by neozenith** [comment]
- Ok looks like Python3.8 CI is passing now there is official support on TravisCI.
 - The reported aiohttp bugs are attracting no PRs to address them.
 - The other fixes have been in the wild for a few months now.

I'm looking to close this PR by the end of Nov 2019 pending time taken to get #478 merged and released as a minor release.

**Comment by neozenith** [comment]
Ok it is finally happening! I grafted the commits from this branch onto branch `v4.0.x` in PR#499.

Thanks for your work on this PR @hugovk and your immense patience and continual rebasing.

Closing this in favor of #499

**Comment by hugovk** [comment]
You're welcome, thanks for maintaining VCR.py!

**Pull request closed without merging**